### PR TITLE
core: remove proxy when cleared in all cases. resolves #7411

### LINKED ISSUE #7411: Proxy settings not being removed until manual restart of tray or container
"So I started up a sandbox and ran a fresh copy of the Windows binaries for Jackett in it [...and added a NordVPN server for the proxy. It worked as it should].

I removed the proxy [...], applied and let it refresh, tested again [...but it behaved as if the proxy settings were still applied]. It wasn't until I shutdown the tray program and restarted it that [...it behaved as if the proxy settings were removed].

To test this in my actual Docker Jackett install I removed the proxy, applied, let it refresh, tested, [...and it behaved as if the proxy settings were still applied]. Applying server settings multiple times didn't help. So I restarted the container via Portainer, and only then did [...it behave as if the proxy settings were removed].

Whatever is happening during that refresh isn't actually removing the proxy settings. I've confirmed this by successfully running tests on indexers which should fail without the proxy, [and having a test fail which should succeed - BadassTorrents,] both in the Docker container and the sandboxed Windows install."

https://github.com/Jackett/Jackett/issues/7325#issuecomment-591221455

### PR DESCRIPTION
**Pull request by cadatoiva**
Also implements "Don't Repeat Yourself" (DRY) principles by moving duplicated proxy related static functions/vars into parent class.

Fixes #7411

### CODE REVIEW

**Comment by ngosang** [comment]
I will look at this after the test pass but I want to ask @Jackett/maintainers 
Why we have 4 different implementations of WebClient? I would understand having two if there is a bug in mono or something like that, but four?

**Comment by cadatoiva** [comment]
There's technically just 2, `HttpWebClient` and `HttpWebClient2` and then there's a `*NetCore` version for both because dotnet-core requires different imports/types to work from what I saw of the code. My FreeBSD install ends up using `HttpWebClient` because Mono + *2 is incompatible(?). But my Windows test environment uses `HttpWebClient2*` depending on whether I test net461 or dotnet-core builds.

I'd need to dig deeper into the code than I did during this to see why there's a v1 and v2, or if we can't do something like `#if net461 ... #else ... #endif` for 1 or two lines to absorb the NetCore versions.

**Comment by ngosang** [comment]
This is the core of Jackett. I would like to take my time to review this in detail, maybe this weekend.

Meanwhile, if you have time, I think we can get big gains if we understand, simplify and refactor the WebClient. For example how many clients are we creating? one per connection? per tracker? can we have a pool with 10 clients and reuse them all the time?

**Comment by ngosang** [comment]
You can configure any of them => https://github.com/Jackett/Jackett/blob/master/src/Jackett.Server/Initialisation.cs#L14
But in the practice, only 2 are used => https://github.com/Jackett/Jackett/blob/master/src/Jackett.Server/Program.cs#L52

@garfield69 can you help us?

**Comment by garfield69** [comment]
I don't know much about the inner workings of the clients, other than what  is already in the source comments right at the start of httpWebClient2
```
    // Compared to HttpWebClient this implementation will reuse the HttpClient instance (one per indexer).
    // This should improve performance and avoid problems with too man open file handles.
```
and HttpWebClientNetCore and HttpWebClient2NetCore came later, and are just clones of the first two clients altered for netcore.
and that is pretty much all I know about them.

**Comment by cadatoiva** [comment]
I started to refractor the abstract class, but I'll break it all down and see if we can't simplify this to a single client, especially since we're only using two of them.

**Comment by cadatoiva** [comment]
Ok, so I'm elbows deep into this section of the code, and have been working on it for going on 5 hours now. Here's what I've found:

The difference between * & *NetCore was just the idea that `X509Certificate2` and/or where the `ValidateCertificate` callback function gets registered was a NetCore only feature. This is wrong. Without that, the difference between these two class of versions is <10% of their respective code bases.

The full 8 lines of `GetBytes` are duplicated in `GetString` and in both functions we collect a `WebClientByteResult`. The only difference is that once we have the byte result in `GetString` we do some extra processing to find the character encoding and convert the `Byte[]` into a string.

As such I think we should merge the 3 web result classes into a single one, and have a `ContentString` and `ContentBytes` access point that does this conversion internal to the class. This will also help us move away from using AutoFac mapping like we've already discussed by removing half of its usage.

And finally, we've been defaulting to HttpClient2NetCore for a while in NetCore installations, and it's supposed to be more performant. So it seems like migrating to the v2 Client, and merging this all into the base class as a single file should be the move to make.

I have more testing to do, figuring out if that last 10% are real differences or just the fact that we've been maintaining 4 nearly identical copies of a single class and not patching all 4 at the same time.

**Comment by cadatoiva** [comment]
This is what I have so far on merging the clients. The *NetCore versions are gone, and most of the code is merged into the base WebClient class. There's still some bugs in a couple of trackers right now that I'm hunting down. Specifically HorribleSubs and Torrentz2 I know aren't working when you run Test. The error has been `NullReferenceException` but the trace is off for some reason, making it hard to find.

File: src/Jackett.Common/Indexers/CardigannIndexer.cs [hunk 1]
```diff
@@ -764,9 +764,7 @@ protected string getRedirectDomainHint(string requestUrl, string RedirectUrl)
             return null;
         }
 
-        protected string getRedirectDomainHint(WebClientByteResult result) => getRedirectDomainHint(result.Request.Url, result.RedirectingTo);
-
-        protected string getRedirectDomainHint(WebClientStringResult result) => getRedirectDomainHint(result.Request.Url, result.RedirectingTo);
+        protected string getRedirectDomainHint(WebResult result) => getRedirectDomainHint(result.Request.Url, result.RedirectingTo);
```
Comment (cadatoiva) [inline src/Jackett.Common/Indexers/CardigannIndexer.cs]: In retrospect, changes like this should have been made even before merging the two classes. These properties are in the base class, and there was no reason to duplicate the function for the two derived classes.

**Comment by ngosang** [comment]
@cadatoiva I can't review this PR, it contains too many changes. Please open several small PRs.
For example:
* https://github.com/Jackett/Jackett/pull/7631#discussion_r392561808
* Merge classes
* Rename `ContentString` and other methods
* Fix the proxy thing

**Comment by cadatoiva** [comment]
I mean, I can try breaking it up into smaller PRs, but ~90 out of the 98 files changed are just the property/function names being changed in the indexers. GitHub does have features to help manage the review process, such as letting you view specific commits and marking files as viewed to collapse them: 
image

**Comment by ngosang** [comment]
> I mean, I can try breaking it up into smaller PRs, but ~90 out of the 98 files changed are just the property/function names being changed in the indexers.

Do it please. I don't want to just review the code, I want to test it too and there are too many thing in this PR. Let this open as reference.

**Comment by cadatoiva** [comment]
I've made a feature branch in the Jackett org to work on the changes piecemeal as you've requested. What I had so far was buggy as is, so maybe this will help catch the bugs before they exist anyway.

This PR is now solely dedicated to fixing the bug with the proxy as outlined in the issue attached to the PR.

**Pull request merged**